translator/metrics/otlp_to_oc.go

```diff
@@ -0,0 +1,377 @@
+		LabelValues:    labelsToOC(point.Labels, labelKeys),
+		Points: []*ocmetrics.Point{
+			{
+				Timestamp: unixnanoToTimestamp(point.TimestampUnixnano),
+				Value: &ocmetrics.Point_DoubleValue{
+					DoubleValue: point.Value,
+				},
+			},
+		},
+	}
+}
+
+func histogramPointToOC(point *otlpmetrics.HistogramDataPoint, labelKeys []*ocmetrics.LabelKey) *ocmetrics.TimeSeries {
+	return &ocmetrics.TimeSeries{
+		StartTimestamp: unixnanoToTimestamp(point.StartTimeUnixnano),
+		LabelValues:    labelsToOC(point.Labels, labelKeys),
+		Points: []*ocmetrics.Point{
+			{
+				Timestamp: unixnanoToTimestamp(point.TimestampUnixnano),
+				Value: &ocmetrics.Point_DistributionValue{
+					DistributionValue: &ocmetrics.DistributionValue{
+						Count:                 int64(point.Count),
+						Sum:                   point.Sum,
+						SumOfSquaredDeviation: 0,
+						BucketOptions:         histogramExplicitBoundsToOC(point.ExplicitBounds),
+						Buckets:               histogramBucketsToOC(point.Buckets),
+					},
+				},
+			},
+		},
+	}
+}
+
+func histogramExplicitBoundsToOC(bounds []float64) *ocmetrics.DistributionValue_BucketOptions {
+	if len(bounds) == 0 {
+		return nil
+	}
+
+	return &ocmetrics.DistributionValue_BucketOptions{
+		Type: &ocmetrics.DistributionValue_BucketOptions_Explicit_{
+			Explicit: &ocmetrics.DistributionValue_BucketOptions_Explicit{
+				Bounds: bounds,
+			},
+		},
+	}
+}
+
+func histogramBucketsToOC(buckets []*otlpmetrics.HistogramDataPoint_Bucket) []*ocmetrics.DistributionValue_Bucket {
+	if len(buckets) == 0 {
+		return nil
+	}
+
+	ocBuckets := make([]*ocmetrics.DistributionValue_Bucket, 0, len(buckets))
+	for _, bucket := range buckets {
+		ocBuckets = append(ocBuckets, &ocmetrics.DistributionValue_Bucket{
+			Count:    int64(bucket.Count),
+			Exemplar: exemplarToOC(bucket.Exemplar),
+		})
+	}
+	return ocBuckets
+}
+
+func exemplarToOC(exemplar *otlpmetrics.HistogramDataPoint_Bucket_Exemplar) *ocmetrics.DistributionValue_Exemplar {
+	if exemplar == nil {
+		return nil
+	}
+
+	return &ocmetrics.DistributionValue_Exemplar{
+		Value:       exemplar.Value,
+		Timestamp:   unixnanoToTimestamp(exemplar.TimestampUnixnano),
+		Attachments: exemplarAttachmentsToOC(exemplar.Attachments),
+	}
+}
+
+func exemplarAttachmentsToOC(attachments []*otlpcommon.StringKeyValue) map[string]string {
+	if len(attachments) == 0 {
+		return nil
+	}
+
+	ocAttachments := make(map[string]string, len(attachments))
+	for _, att := range attachments {
+		ocAttachments[att.Key] = att.Value
+	}
+	return ocAttachments
+}
+
+func labelsToOC(labels []*otlpcommon.StringKeyValue, labelKeys []*ocmetrics.LabelKey) []*ocmetrics.LabelValue {
+	if len(labels) == 0 {
+		return nil
+	}
+
+	// NOTE: We need to set label values in the same order as keys, thus
+	// intermediate transformation to a map for fast lookups
+	labelMap := make(map[string]string, len(labels))
+	for _, label := range labels {
+		labelMap[label.Key] = label.Value
+	}
```

Comment:
I think we can do better. Building this map is not necessary. Instead we can do this:

1. Make `allLabelKeys` return both the slice of `labelKeys` and a map of label keys to index in the slice (let's name `labelKeysIndices`). `labelKeysIndices` can replace `uniqueKeys` if we declare it as `map[string]int`.
2. Here in labelsToOC create a slice of `ocmetrics.LabelValue` and set all `HasValue=false`.
3. Iterate over `labels` and for each label that is in `labelKeysIndices` set `HasValue=true` and the value of the label at the corresponding index.

Benefit: no need to build a map in this func and we just reuse already built map from `allLabelKeys`.

I think this should work and should be faster.

Reply:
Agreed, that seems like a reasonable approach.